app/src/behaviors/behavior_macro.c

```diff
@@ -0,0 +1,134 @@
+/*
+ * Copyright (c) 2020 The ZMK Contributors
+ *
+ * SPDX-License-Identifier: MIT
+ */
+
+#include <device.h>
+#include <drivers/behavior.h>
+#include <logging/log.h>
+#include <zmk/behavior.h>
+
+#include <zmk/matrix.h>
+#include <zmk/endpoints.h>
+#include <zmk/event_manager.h>
+#include <zmk/events/keycode_state_changed.h>
+#include <zmk/events/modifiers_state_changed.h>
+#include <zmk/hid.h>
+
+#define DT_DRV_COMPAT zmk_behavior_macro
+
+#if DT_HAS_COMPAT_STATUS_OKAY(DT_DRV_COMPAT)
+
+LOG_MODULE_DECLARE(zmk, CONFIG_ZMK_LOG_LEVEL);
+
+enum mode {
+    MODE_KEY_UP = 0,
+    MODE_KEY_DOWN = 1,
+    MODE_HOLD = 2,
+};
+
+struct behavior_macro_config {
+    int mode;
+    int sleep;
+    int behavior_count;
+    struct zmk_behavior_binding *behaviors;
+};
+
+static int behavior_macro_init(const struct device *dev) { return 0; }
+
+static void press_and_release_combo(const struct behavior_macro_config *cfg,
+                                    struct zmk_behavior_binding_event event) {
+    for (int index = 0; index < cfg->behavior_count; index++) {
+        const struct device *behavior = device_get_binding(cfg->behaviors[index].behavior_dev);
+        if (!behavior) {
+            break;
+        }
+        behavior_keymap_binding_pressed(&cfg->behaviors[index], event);
+        k_msleep(cfg->sleep);
+        behavior_keymap_binding_released(&cfg->behaviors[index], event);
+        k_msleep(cfg->sleep);
+    }
+}
+
+static void press_combo(const struct behavior_macro_config *cfg,
+                        struct zmk_behavior_binding_event event) {
+    for (int index = 0; index < cfg->behavior_count; index++) {
+        const struct device *behavior = device_get_binding(cfg->behaviors[index].behavior_dev);
+        if (!behavior) {
+            break;
+        }
+        behavior_keymap_binding_pressed(&cfg->behaviors[index], event);
+        k_msleep(cfg->sleep);
+    }
+}
+
+static void release_combo(const struct behavior_macro_config *cfg,
+                          struct zmk_behavior_binding_event event) {
+    for (int index = 0; index < cfg->behavior_count; index++) {
+        const struct device *behavior = device_get_binding(cfg->behaviors[index].behavior_dev);
+        if (!behavior) {
+            break;
+        }
+        behavior_keymap_binding_released(&cfg->behaviors[index], event);
+        k_msleep(cfg->sleep);
+    }
+}
+
+static int on_keymap_binding_pressed(struct zmk_behavior_binding *binding,
+                                     struct zmk_behavior_binding_event event) {
+    const struct device *dev = device_get_binding(binding->behavior_dev);
+    const struct behavior_macro_config *cfg = dev->config;
+    if (cfg->mode == MODE_KEY_DOWN) {
```

Comment:
When I was first reviewing this, it was difficult for me to understand what it was doing and how it worked. I think that is mainly because of the way it does a completely different thing between hold and up/down modes.

It feels like these are conceptually two different things, and the only code they really share is the DT binding handling. I wonder if it would make more sense to split them into separate behaviors?

1. macro: Press and release a bunch of things in sequence when I press a key (with option to trigger on release instead)
2. multi(?): Press a bunch of things (optionally staggered with a delay) and hold them until I release the key, then release them in the same order

https://github.com/zmkfirmware/zmk/issues/689 seems like it would extend/replace the first but not the second.